MoreLinq.Test/TrySingleTest.cs

```diff
@@ -0,0 +1,193 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 James Webster. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq.Test
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class TrySingleTest
+    {
+        [TestCase(SourceKind.Sequence)]
+        [TestCase(SourceKind.BreakingList)]
+        [TestCase(SourceKind.BreakingReadOnlyList)]
+        [TestCase(SourceKind.BreakingCollection)]
+        [TestCase(SourceKind.BreakingReadOnlyCollection)]
+        public void TrySingleWithEmptySource(SourceKind kind)
+        {
+            var arrayWithNone = new int?[0].ToSourceKind(kind);
+
+            var (cardinality, value) = arrayWithNone.TrySingle("zero", "one", "many");
+
+            Assert.AreEqual(cardinality, "zero");
+            Assert.IsNull(value);
+        }
+
+        [TestCase(SourceKind.Sequence)]
+        [TestCase(SourceKind.BreakingList)]
+        [TestCase(SourceKind.BreakingReadOnlyList)]
+        public void TrySingleWithSingleton(SourceKind kind)
+        {
+            var arrayWithOne = new int?[] { 10 }.ToSourceKind(kind);
+
+            var (cardinality, value) = arrayWithOne.TrySingle("zero", "one", "many");
+
+            Assert.AreEqual(cardinality, "one");
+            Assert.AreEqual(value, 10);
+        }
+
+        [TestCaseSource(nameof(_singletonCollectionTestCases))]
+        public void TrySingleWithSingletonCollections<T>(IEnumerable<T> collection, T result)
+        {
+            var (cardinality, value) = collection.TrySingle("zero", "one", "many");
+
+            Assert.AreEqual(cardinality, "one");
+            Assert.AreEqual(value, result);
+        }
+
+        private static object[] _singletonCollectionTestCases =
+        {
+            new object[] { new BreakingSingleElementCollection<int>(10), 10 },
```

Comment:
Given above, this together with the next should instead use instantiations of `TestCaseData`:

```c#
            new TestCaseData(new BreakingSingleElementCollection<int>(10), 10),
            new TestCaseData(new BreakingSingleElementReadOnlyCollection<int>(20), 20),
```